OSFUSE-236: support for fat-jar applications

nicolaferraro opened the pull request:
I changed the s2i java image making it able to recognize hawtapp builds from fat-jar builds and run the image accordingly.

rhuss commented on java/s2i/assemble:
```diff
@@ -93,17 +114,25 @@ function build_maven() {
-  echo "Copying Maven artifacts from ${build_dir}/hawt-app to ${app_dir} ..."
-  copy_dir "${build_dir}/hawt-app" ${app_dir}
+  if [ -d "${build_dir}/hawt-app" ]; then
+    echo "Hawtapp plugin detected. Copying Maven artifacts from ${build_dir}/hawt-app to ${app_dir} ..."
```
This works only when hawt-app is configured in the pom.xml, right ? The previous version worked also with hawt-app when only a JAVA_MAIN_CLASS is given and no configuration of hawt-app within the pom.xml (not that I say that was a good solution, just saying /wrt to backwards compatibility ;-)

nicolaferraro replied:
Yes, I know.. but it forced the execution of the hawtapp maven goal, that is something you don't desire in non-hawtapp application. I considered that most of the quickstarts and samples have the hawtapp plugin configured, so there is no need to call that goal in the image builder.

Also, you can still use hawtapp by providing specific `MAVEN_ARGS` in the template, even if the pluign is not configured in the pom. The new image just checks the output of the maven build.

rhuss replied:
No problem for me, just wanted to be sure.

rhuss commented on java/s2i/run:
```diff
@@ -10,4 +10,14 @@ JAVA_OPTIONS="${JAVA_OPTIONS:+${JAVA_OPTIONS} }$(/opt/jolokia/jolokia-opts)"
+    exec ${DEPLOYMENTS_DIR}/bin/run.sh
+elif [ -f "${DEPLOYMENTS_DIR}/run-java.sh" ]; then
+    echo "Starting Java Application ..."
+    exec ${DEPLOYMENTS_DIR}/run-java.sh
```
run-java.sh can cope with hawt-app app structure as well, btw (i.e. using a `classpath` file for creating the classpath).

rhuss commented:
Thanks for PR ! 

Nevermind the failing test (which should not run for an PR because it needs specific credentials). Going to fix this now.

I wonder whether we can make it more elegant without copying things around. On my todo list was a unifications of both images for the common parts and refering to run-java-sh and jolokia to the external repositories directly with [fish-pepper](https://github.com/fabric8io-images/fish-pepper) (much like https://github.com/fabric8io-images/java does). That would avoid copying of the script but refering directly to the original script.

The other idea was to get rid of hawt-app alltogether since the running flat classpath apps with a defined classpath ordering is also possible without hawt-app. Also, a single run.sh could be used, making maintenance easier. The only thing one would loose is an hawtapp tar for running it standalone without any Docker env (dont think its that crucial, though).

nicolaferraro commented:
Yes, I think it is a good idea to use fish-pepper as done for the other java images. I made just few modifications because I've seen the repos were so distant... Using a single run.sh is also a good idea, I see that its [run.sh](https://github.com/fabric8io/fabric8/blob/master/hawt-app-maven-plugin/src/main/resources/io/fabric8/maven/hawt/app/bin/run.sh) is not doing more than the new run-java-sh.

But I don't get how you can get rid of hawtapp.. I mean, not without breaking compatibility with eg. projects generated using old archetypes.

rhuss commented:
The idea is that for project using already hawt-app, run-java.sh picks up the hawt-app generated class path, ideally with only backward based references to hawt-app specific pathes in the assemble script.
The s2i should smoothly run with existing hawtapp based apps (but this is not really hard if run-java.sh supports hawtapp style apps).

What I mean with 'get rid of' is that in the next versions of the archetypes, hawt-app could be omitted. The generation of the classpath file could be done either by the maven-dependency-plugin (which hawt-app utilizes internally anyway) or even by an extension for the docker-maven-plugin (which will be part of the fabric8-maven-plugin V3) which writes out the classpath files into the assembly. 

To review the actual benefits of hawt-app:

* Providing a `classpath` file for starting flat classpath apps. Can be done with other means, too (dependency-maven-plugin, or by f-m-p via a configuration option (tbd))
* Providing a `run.sh` extracted from the plugin. To be replaced with a run.sh from the base image.
* Creating a `hawtapp.tar` for standalone usage without Docker. There is no substitute for this use case.

The main motivation for getting rid of hawt-app is to boil down the pom.xml for quickstarts, so going down from 3 plugins (hawt-app, d-m-p and f-m-p), to a single one (f-m-p) with some strong defaults (e.g for the base image). We already have migrated quite some of the quickstarts to https://github.com/fabric8-quickstarts (for example https://github.com/fabric8-quickstarts/cdi-cxf)

nicolaferraro commented:
Another task currently performed by hawtapp is identifying the main class. It is usually contained in the maven pom.xml and written to the `run.sh` script via a `${hawtapp.mvn.main.property}`. I'm looking for a clean way to take is during the assemble phase and put it eg. in the `setenv.sh`...

nicolaferraro commented:
I changed the configuration of the java image to be built with fish-pepper (now the `fish-pepper make` command produces the two dockerfiles of the old build, even if the rhel one seems to have wrong references and cannot be built).

I added an option to the assemble script: `ARTIFACT_COPY_ARGS`. It allows the user to specify the artifact that should be copied to the application dir. Defaults to `-r hawt-app/*` if the `hawt-app` dir is present, otherwise `*.jar`, making the app dir ready for the `java-run-sh` script.
This property allows getting rid of hawtapp, as also projects that use the `copy-dependencies` maven goal can be built and run with S2I.

The `java-run-sh` is copied under `/opt/java-run` and launched from there (this fixes an issue when using the image with the f8 build workflow).

The image does not use the `java-run-sh` if the build provides its own `bin/run.sh` script. This is of course for hawt-app compatibility, because the plugin encodes the main class into the `run.sh` script and I cannot find a clean way to extract it (I feel it's better using it directly, anyway, I can change it... don't know how).

I did not include `agent-bond`, as it is better stick with jolokia only for now.

If it is ok, I can fish-pepper the other image as well. I don't think is a good idea making a unique image, they are so different, also from a user's viewpoint.

rhuss commented:
Awesome ! 

I will have a look soon and will give feedback. My hope was to even unify the karaf image but if its too different then its not worth the effort (infact @dhirajsb and I worked on them separately), maybe when rewriting the karaf assemble / run script it could make sense.

Agreed not to include agent-bond, but there's also a fish-pepper block for Jolokia only available --> https://github.com/fabric8io/agent-bond/tree/master/fish-pepper/jolokia (stored in the agent-bond repo, though).

Give me a bit, need to do some other work today, but will be back soon.

Thanks!

nicolaferraro commented:
No worries ;)

rhuss commented:
LGTM, except that maybe we should fine tune the block handling. (see comments in #21). If you dont mind I would work on this tomorrow because I need to refresh my memories also how to do this ;-)

nicolaferraro commented:
No problems at all. Thank you!

rhuss commented:
Added you changes in a slightly different way with https://github.com/fabric8io-images/s2i/commit/c6d9e21e8c33708d820e99c97f88be9557af870b
